py/dml/types_test.py

```diff
@@ -0,0 +1,282 @@
+
+        # bitfields
+        self.assert_neq(TInt(8, False), TInt(8, False, members={}))
+
+        def bitfields():
+            return TInt(32, False, { 'a': (TInt(8, False), 13, 6),
+                                     'b': (TInt(13, False), 27, 15) })
+
+        self.assert_eq(TInt(8, False, members={}), TInt(8, False, members={}))
+        self.assert_eq(bitfields(), bitfields())
+
+        t = bitfields()
+
+        def neq_bitfields_testcase():
+            nonlocal t
+            self.assert_neq(bitfields(), t)
+            t = bitfields()
+
+        # Presence of members matter
+        del t.members['a']
+        neq_bitfields_testcase()
+        # type of members matters
+        t.members['a'][0].signed = True
+        neq_bitfields_testcase()
+        # msb/lsb of members matters
+        t.members['a'] = (TInt(8, False), 12, 5)
+        neq_bitfields_testcase()
+        # order matters
+        t.members = dict(reversed(t.members.items()))
+        neq_bitfields_testcase()
+        # names matter
+        t.members['c'] = t.members['b']
+        del t.members['b']
+        neq_bitfields_testcase()
+
+    def test_TEndianInt(self):
+        self.assert_eq(TEndianInt(8, False, 'big-endian'),
+                       TEndianInt(8, False, 'big-endian'))
+        self.assert_neq(TEndianInt(8, False, 'big-endian'),
+                        TEndianInt(8, False, 'little-endian'))
+
+    def test_TFloat(self):
+        self.assert_eq(TFloat('double'), TFloat('double'))
+        self.assert_neq(TFloat('double'), TFloat('float'))
+
+    def mkTArray(self, size, signed=False):
+        size = (ctree.mkIntegerConstant(None, size, signed) if size is not None
+                else expr.mkLit(None, 'lit', TInt(32, False)))
+        return TArray(TInt(32, False), size)
+
+    def test_TArray(self):
+        # Arrays of constant size need the sizes be equal to be equal.
+        # The type of the size expression is irrelevant.
+        self.assert_eq(self.mkTArray(4), self.mkTArray(4))
+        self.assert_eq(self.mkTArray(4), self.mkTArray(4, True))
+        self.assert_neq(self.mkTArray(4), self.mkTArray(5))
+
+        # Arrays of non-constant size needs the underlying expression to be
+        # identical in order to be equal.
+        self.assert_neq(self.mkTArray(None), self.mkTArray(None))
+        a = self.mkTArray(None)
+        b = self.mkTArray(None)
+        b.size = a.size
+        self.assert_eq(a, b)
+
+        a = self.mkTArray(4)
+        b = self.mkTArray(4)
+
+        def array_testcase(res):
+            nonlocal a, b
+            if res:
+                self.assert_eq(a, b)
+            else:
+                self.assert_neq(a, b)
+
+            a = self.mkTArray(4)
+            b = self.mkTArray(4)
+
+        # Base type matters
+        a.base.signed = True
+        array_testcase(False)
+
+        # Constness: constness of base type and array are normalized
+        a.const = True
+        array_testcase(False)
```

Comment:
Same here, what about:
```
def two_array_types():
  return (self.mkTArray(4), self.mkTArray(4))
...
self.assert_eq(a, b)
(a, b) = two_array_types()
```

Reply:
dun wanna

(for a more persuasive answer, see above)

Reply:
context manager used instead